src/message/dkim.rs

```diff
@@ -0,0 +1,318 @@
+            format!("{}\r\n", re.replace_all(&value, " ").to_string().trim_end())
+        }
+    }
+}
+
+/// Canonicalize signed headers passed as headers_list among mail_headers using canonicalization
+fn dkim_canonicalize_headers(
+    headers_list: Vec<String>,
+    mail_headers: &Headers,
+    canonicalization: DkimCanonicalizationType,
+) -> String {
+    let mut signed_headers = Headers::new();
+    let mut signed_headers_relaxed = String::new();
+    for h in headers_list.into_iter() {
+        let h = match canonicalization {
+            DkimCanonicalizationType::Simple => h,
+            DkimCanonicalizationType::Relaxed => {
+                let mut ret = String::new();
+                ret.push_str(&h);
+                ret.to_lowercase()
+            }
+        };
+        if let Some(value) = mail_headers.get_raw(&h) {
+            match canonicalization {
+                DkimCanonicalizationType::Simple => signed_headers.append_raw(
+                    HeaderName::new_from_ascii(h).unwrap(),
+                    dkim_canonicalize_header_value(value, canonicalization),
+                ),
+                DkimCanonicalizationType::Relaxed => signed_headers_relaxed.push_str(&format!(
+                    "{}:{}",
+                    h,
+                    dkim_canonicalize_header_value(value, canonicalization)
+                )),
+            }
+        }
+    }
+    match canonicalization {
+        DkimCanonicalizationType::Simple => format!("{}", signed_headers),
+        DkimCanonicalizationType::Relaxed => signed_headers_relaxed,
+    }
+}
+
+/// Sign with Dkim a message by adding Dkim-Signture header created with configuration expressed by
+/// dkim_config
+///
+
+pub fn dkim_sign(message: &mut Message, dkim_config: DkimConfig) {
+    let timestamp = format!(
+        "{}",
+        SystemTime::now()
+            .duration_since(SystemTime::UNIX_EPOCH)
+            .unwrap()
+            .as_secs()
+    );
+    let headers = message.headers();
+    let body_hash = Sha256::digest(
+        dkim_canonicalize_body(&message.body_raw(), dkim_config.canonicalization.body).as_bytes(),
+    );
+    let bh = encode(body_hash);
+    let signed_headers_list = match dkim_config.canonicalization.header {
+        DkimCanonicalizationType::Simple => dkim_config.headers.join(":"),
+        DkimCanonicalizationType::Relaxed => dkim_config.headers.join(":").to_lowercase(),
+    };
+    let dkim_header = dkim_header_format(
+        dkim_config.domain.clone(),
+        dkim_config.selector.clone(),
+        dkim_config.canonicalization,
+        timestamp.clone(),
+        signed_headers_list.clone(),
+        bh.clone(),
+        "".to_string(),
+    );
+    let signed_headers = dkim_canonicalize_headers(
+        dkim_config.headers.clone(),
+        headers,
+        dkim_config.canonicalization.header,
+    );
+    let private_key = RsaPrivateKey::from_pkcs1_pem(&dkim_config.private_key).unwrap();
```

Comment:
This parsing should probably be done when building the dkim_config then just store the parsed version. This uses less memory and moves the errors to something that is easier and more likely to run at startup.